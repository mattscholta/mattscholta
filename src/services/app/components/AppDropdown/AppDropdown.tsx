import * as React from 'react';
import classnames from 'classnames';

import { isMobileDevice } from '@src/utils/device';

import styles from './styles.module.css';

export interface AppDropdownProps {
  content: any;
  dropdown: any;
}

/**
 * @name AppDropdown
 * @description A simple and reusable Dropdown component that takes the
 * target "content" and a "dropdown" which is rendered as the users
 * hovers or otherwise activates the "content area"
 */
export const AppDropdown = (props: AppDropdownProps) => {
  const { content: Content, dropdown: Dropdown } = props;

  // Hooks
  const [hover, setHover] = React.useState(false);

  // Styles
  const tailwind = 'u-flex u-h-full u-items-stretch u-relative';
  const cssComponent = classnames(tailwind, styles.component);
  const cssContainer = classnames(styles.container, { [styles.close]: !hover });

  // Handlers
  const onBlur = () => {
    if (hover) setHover(false);
  };

  const onClick = () => {
    setHover(!hover);
  };

  const onHover = () => {
    const isMobile = isMobileDevice();

    if (isMobile) return;
    if (!hover) setHover(true);
  };

  return (
    <div
      className={cssComponent}
      onBlur={onBlur}
      onClick={onClick}
      onFocus={onHover}
      onMouseEnter={onHover}
      onMouseLeave={onBlur}
    >
      <Content />
      <div className={cssContainer}>
        <Dropdown className={styles.dropdown} />
      </div>
    </div>
  );
};
