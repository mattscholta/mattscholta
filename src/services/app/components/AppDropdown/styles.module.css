.component {
  cursor: pointer;
  position: relative;
}

.container {
  box-shadow: 0 4px 4px rgba(0, 0, 0, 0.15);
  position: absolute;
  right: 0;
  top: 100%;
  transform: translateX(0%);
  transition: opacity 140ms ease-in-out;
  z-index: 2;
}

.dropdown {
  background-color: theme('colors.white');
  border: 1px solid theme('colors.forest');
  cursor: default;
  opacity: 1;
  padding: theme('spacing.2x');
}

.container.close {
  opacity: 0;
  pointer-events: none;
  z-index: 1;
}

/* Only on large screes can it sit in the middle */
@media (min-width: 1568px) {
  .container {
    right: 50%;
    transform: translateX(50%);
  }
}
