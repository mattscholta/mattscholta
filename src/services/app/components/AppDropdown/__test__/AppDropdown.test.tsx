import * as React from 'react';
import { render, RenderResult } from '@testing-library/react';

import { AppDropdown, AppDropdownProps } from '../AppDropdown';

describe('AppDropdown', () => {
  const Content = jest.fn().mockImplementation(() => null);
  const Dropdown = jest.fn().mockImplementation(() => null);

  let props: AppDropdownProps;
  let wrapper: RenderResult;

  const createWrapper = () => <AppDropdown {...props} />;

  beforeEach(() => {
    props = {
      content: Content,
      dropdown: Dropdown
    };

    wrapper = render(createWrapper());
  });

  it('matches snapshot', () => {
    expect(wrapper.container.childNodes).toMatchSnapshot();
  });
});
