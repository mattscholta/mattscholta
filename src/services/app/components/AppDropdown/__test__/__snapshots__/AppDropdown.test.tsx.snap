// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppDropdown matches snapshot 1`] = `
NodeList [
  <div
    class="u-flex u-h-full u-items-stretch u-relative component"
  >
    <div
      class="container close"
    />
  </div>,
]
`;
